File: android/app/src/test/kotlin/com/aispotlight/android/core/DocumentContractTest.kt
package com.aispotlight.android.core

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/** The Kotlin twin of scripts/DocumentContractTest.swift — same cases. */
class DocumentContractTest {
    private val mb = 1024L * 1024

    @Test fun preflightRules() {
        assertEquals(DocumentPreflight.Verdict.Accepted,
            DocumentPreflight.check("pdf", 10 * mb, false, 0, 0))
        assertEquals(DocumentPreflight.Verdict.Accepted,
            DocumentPreflight.check("DOCX", 100, false, 2, 0))
        assertEquals(DocumentPreflight.Verdict.TooManyDocuments(3),
            DocumentPreflight.check("docx", 100, false, 3, 0))
        assertEquals(DocumentPreflight.Verdict.UnsupportedType("exe"),
            DocumentPreflight.check("exe", 100, false, 0, 0))
        assertEquals(DocumentPreflight.Verdict.UnsupportedType("png"),
            DocumentPreflight.check("png", 100, false, 0, 0))
        assertEquals(DocumentPreflight.Verdict.EmptyFile,
            DocumentPreflight.check("pdf", 0, false, 0, 0))
        assertEquals(DocumentPreflight.Verdict.Encrypted,
            DocumentPreflight.check("pdf", 100, true, 0, 0))
        assertEquals(DocumentPreflight.Verdict.FileTooLarge(50 * mb),
            DocumentPreflight.check("pdf", 51 * mb, false, 0, 0))
        assertEquals(DocumentPreflight.Verdict.MessageTooLarge(50 * mb),
            DocumentPreflight.check("pdf", 30 * mb, false, 1, 25 * mb))
        assertTrue(DocumentPreflight.isDocumentMime("application/pdf"))
        assertFalse(DocumentPreflight.isDocumentMime("image/png"))
        assertTrue(DocumentPreflight.isLocallyReadable("application/pdf"))
        assertFalse(DocumentPreflight.isLocallyReadable(DocumentPreflight.mimeType("xlsx")!!))
        assertEquals("application/pdf", DocumentPreflight.documentMime("a.PDF", "application/octet-stream"))
        assertNull(DocumentPreflight.documentMime("a.exe", "application/octet-stream"))
    }

    private val joined = DocumentTextQuery.join(listOf("First page text.", "", "Third page.\n\nSecond paragraph."))

    @Test fun pageMarkersRoundTrip() {
        val pages = DocumentTextQuery.pages(joined)
        assertEquals(3, pages.size)
        assertEquals("First page text.", pages[0])
        assertEquals("", pages[1])
        assertTrue(pages[2].startsWith("Third page."))
        assertEquals(1, DocumentTextQuery.pages("plain text without markers").size)
    }

    @Test fun pageRanges() {
        assertEquals(3..5, DocumentTextQuery.parsePageRange("3-5", 10))
        assertEquals(7..7, DocumentTextQuery.parsePageRange("7", 10))
        assertEquals(2..4, DocumentTextQuery.parsePageRange(" 2 – 4 ", 10))
        assertEquals(8..10, DocumentTextQuery.parsePageRange("8-20", 10))
        assertNull(DocumentTextQuery.parsePageRange("12", 10))
        assertNull(DocumentTextQuery.parsePageRange("abc", 10))
        assertNull(DocumentTextQuery.parsePageRange(null, 10))
    }

    @Test fun searchHits() {
        val hits = DocumentTextQuery.search(joined, "PAGE")
        assertEquals(2, hits.size)
        assertEquals(1, hits[0].page)
        assertEquals(3, hits[1].page)
        assertTrue(DocumentTextQuery.search(joined, "   ").isEmpty())
    }

    @Test fun rendering() {
        val whole = DocumentTextQuery.render("a.pdf", joined, null, null)
        assertTrue(whole.startsWith("a.pdf — 3 pages"))
        assertTrue(whole.contains("[Page 3]"))
        val ranged = DocumentTextQuery.render("a.pdf", joined, "3", null)
        assertTrue(ranged.contains("pages 3-3") && ranged.contains("Third page.") && !ranged.contains("First page"))
        assertTrue(DocumentTextQuery.render("a.pdf", joined, "99", null).contains("Invalid page range"))
        val queried = DocumentTextQuery.render("a.pdf", joined, null, "second")
        assertTrue(queried.contains("1 match") && queried.contains("Page 3:"))
        assertTrue(DocumentTextQuery.render("a.pdf", joined, null, "zzz").contains("No matches"))
        val long = "word ".repeat(20_000)
        val capped = DocumentTextQuery.capped(long)
        assertTrue(capped.length <= DocumentTextQuery.MAX_RESULT_CHARACTERS + DocumentTextQuery.TRUNCATION_NOTE.length + 1)
        assertTrue(capped.endsWith(DocumentTextQuery.TRUNCATION_NOTE))
        assertEquals("short", DocumentTextQuery.capped("short"))
    }
}

File: scripts/DocumentContractTest.swift
import Foundation

// Contract test for the document attachment pre-flight and the read_document
// text queries. Compiled STANDALONE with the two pure files — no app target:
//   swiftc DocumentPreflight.swift DocumentTextQuery.swift DocumentContractTest.swift -o test
// Run via scripts/test-attach-note.sh.

nonisolated(unsafe) var failures = 0
func check(_ condition: Bool, _ label: String) {
    if condition { print("  ok   \(label)") } else { failures += 1; print("  FAIL \(label)") }
}

@main
struct DocumentContractTest {
    static func main() {

    print("== pre-flight ==")
    let mb = 1024 * 1024
    check(DocumentPreflight.check(ext: "pdf", bytes: 10 * mb, isEncrypted: false, pendingDocumentCount: 0, pendingDocumentBytes: 0) == .accepted, "pdf accepted")
    check(DocumentPreflight.check(ext: "DOCX", bytes: 100, isEncrypted: false, pendingDocumentCount: 2, pendingDocumentBytes: 0) == .accepted, "extension case-insensitive, third document accepted")
    check(DocumentPreflight.check(ext: "docx", bytes: 100, isEncrypted: false, pendingDocumentCount: 3, pendingDocumentBytes: 0) == .tooManyDocuments(limit: 3), "fourth document refused")
    check(DocumentPreflight.check(ext: "exe", bytes: 100, isEncrypted: false, pendingDocumentCount: 0, pendingDocumentBytes: 0) == .unsupportedType(ext: "exe"), "unsupported type")
    check(DocumentPreflight.check(ext: "png", bytes: 100, isEncrypted: false, pendingDocumentCount: 0, pendingDocumentBytes: 0) == .unsupportedType(ext: "png"), "images are not documents")
    check(DocumentPreflight.check(ext: "pdf", bytes: 0, isEncrypted: false, pendingDocumentCount: 0, pendingDocumentBytes: 0) == .emptyFile, "empty file")
    check(DocumentPreflight.check(ext: "pdf", bytes: 100, isEncrypted: true, pendingDocumentCount: 0, pendingDocumentBytes: 0) == .encrypted, "encrypted pdf")
    check(DocumentPreflight.check(ext: "pdf", bytes: 51 * mb, isEncrypted: false, pendingDocumentCount: 0, pendingDocumentBytes: 0) == .fileTooLarge(limitBytes: 50 * mb), "single file over 50 MB")
    check(DocumentPreflight.check(ext: "pdf", bytes: 30 * mb, isEncrypted: false, pendingDocumentCount: 1, pendingDocumentBytes: 25 * mb) == .messageTooLarge(limitBytes: 50 * mb), "combined over 50 MB")
    check(DocumentPreflight.isDocumentMime("application/pdf") && !DocumentPreflight.isDocumentMime("image/png"), "mime classification")
    check(DocumentPreflight.isLocallyReadable(mime: "application/pdf") && !DocumentPreflight.isLocallyReadable(mime: DocumentPreflight.mimeType(forExtension: "xlsx")!), "local readability")
    check(DocumentPreflight.iconName(forFilename: "a.PDF") == "doc.richtext", "icon by extension")

    print("== page markers ==")
    let joined = DocumentTextQuery.join(pages: ["First page text.", "", "Third page.\n\nSecond paragraph."])
    let pages = DocumentTextQuery.pages(of: joined)
    check(pages.count == 3, "three pages round-trip (empty page kept)")
    check(pages[0] == "First page text." && pages[1] == "" && pages[2].hasPrefix("Third page."), "page contents preserved")
    check(DocumentTextQuery.pages(of: "plain text without markers").count == 1, "unmarked text is one page")

    print("== page ranges ==")
    check(DocumentTextQuery.parsePageRange("3-5", pageCount: 10) == 3...5, "3-5")
    check(DocumentTextQuery.parsePageRange("7", pageCount: 10) == 7...7, "single page")
    check(DocumentTextQuery.parsePageRange(" 2 – 4 ", pageCount: 10) == 2...4, "spaces and en dash")
    check(DocumentTextQuery.parsePageRange("8-20", pageCount: 10) == 8...10, "clamped to the last page")
    check(DocumentTextQuery.parsePageRange("12", pageCount: 10) == nil, "start past the end")
    check(DocumentTextQuery.parsePageRange("abc", pageCount: 10) == nil, "garbage")
    check(DocumentTextQuery.parsePageRange(nil, pageCount: 10) == nil, "nil")

    print("== search ==")
    let hits = DocumentTextQuery.search(joined, query: "PAGE")
    check(hits.count == 2 && hits[0].page == 1 && hits[1].page == 3, "case-insensitive hits with page numbers")
    check(DocumentTextQuery.search(joined, query: "   ").isEmpty, "blank query → no hits")

    print("== rendering ==")
    let whole = DocumentTextQuery.render(name: "a.pdf", text: joined, pageRange: nil, query: nil)
    check(whole.hasPrefix("a.pdf — 3 pages") && whole.contains("[Page 3]"), "whole document with header")
    let ranged = DocumentTextQuery.render(name: "a.pdf", text: joined, pageRange: "3", query: nil)
    check(ranged.contains("pages 3-3") && ranged.contains("Third page.") && !ranged.contains("First page"), "page range")
    check(DocumentTextQuery.render(name: "a.pdf", text: joined, pageRange: "99", query: nil).contains("Invalid page range"), "invalid range message")
    let queried = DocumentTextQuery.render(name: "a.pdf", text: joined, pageRange: nil, query: "second")
    check(queried.contains("1 match") && queried.contains("Page 3:"), "query result")
    check(DocumentTextQuery.render(name: "a.pdf", text: joined, pageRange: nil, query: "zzz").contains("No matches"), "no matches message")
    let long = String(repeating: "word ", count: 20_000)
    let capped = DocumentTextQuery.capped(long)
    check(capped.count <= DocumentTextQuery.maxResultCharacters + DocumentTextQuery.truncationNote.count + 1 && capped.hasSuffix(DocumentTextQuery.truncationNote), "cap with note")
    check(DocumentTextQuery.capped("short") == "short", "short text untouched")

    if failures > 0 {
        print("\(failures) failure(s)")
        exit(1)
    }
    print("all green")
    }
}
